File: ios/LookSeeProto/UI Files/BusinessAnalyticsView.swift
//
//  BusinessAnalyticsView.swift
//  LookSeeProto
//
//  Created by Angel Pineda on 9/9/26.
//


import SwiftUI
import Charts
import Combine
import Foundation

// MARK: - Data Models
struct AnalyticsResponse: Decodable {
    let analytics: [LandmarkAnalyticsData]
}

struct LandmarkAnalyticsData: Decodable, Identifiable {
    var id: String { landmarkId }
    let landmarkId: String
    let label: String
    let totalClicks: Int
    let dailyData: [DailyClickPoint]
}

struct DailyClickPoint: Decodable, Identifiable {
    var id: String { date }
    let date: String
    let clicks: Int
}

// Internal model for clean Time-Series Charts
struct TimeSeriesPoint: Identifiable {
    let id = UUID()
    let date: Date
    let clicks: Int
}

enum AnalyticsChartMode {
    case trend
    case distribution
}

// MARK: - View Model
@MainActor
class BusinessAnalyticsViewModel: ObservableObject {
    @Published var data: [LandmarkAnalyticsData] = []
    @Published var isLoading = true
    @Published var errorMessage: String? = nil

    // Dashboard Aggregations
    var totalViews: Int {
        data.reduce(0) { $0 + $1.totalClicks }
    }
    
    var topLandmark: String {
        guard totalViews > 0 else { return "N/A" }
        return data.max(by: { $0.totalClicks < $1.totalClicks })?.label ?? "N/A"
    }
    
    var activeLandmarksWithViews: [LandmarkAnalyticsData] {
        data.filter { $0.totalClicks > 0 }
            .sorted { $0.totalClicks > $1.totalClicks }
    }
    
    var sortedLandmarks: [LandmarkAnalyticsData] {
        data.sorted {
            if $0.totalClicks == $1.totalClicks {
                return $0.label < $1.label
            }
            return $0.totalClicks > $1.totalClicks
        }
    }
    
    // Converts the string dates ("09/08") into actual Date objects so SwiftUI Charts can draw a proper X-Axis
    var timeSeriesData: [TimeSeriesPoint] {
        guard let template = data.first?.dailyData else { return [] }
        var merged = [String: Int]()
        
        for item in data {
            for point in item.dailyData {
                merged[point.date, default: 0] += point.clicks
            }
        }
        
        let currentYear = Calendar.current.component(.year, from: Date())
        let formatter = DateFormatter()
        formatter.dateFormat = "MM/dd/yyyy"
        
        return template.compactMap { point in
            let dateString = "\(point.date)/\(currentYear)"
            guard let exactDate = formatter.date(from: dateString) else { return nil }
            return TimeSeriesPoint(date: exactDate, clicks: merged[point.date] ?? 0)
        }.sorted { $0.date < $1.date }
    }

    func fetchAnalytics(vm: AuthViewModel) async {
        guard !vm.userId.isEmpty else { return }
        
        guard let url = URL(string: "https://d11vl3v9w133rh.cloudfront.net/analytics") else { return }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        
        let idToken = await vm.fetchIdToken()
        if !idToken.isEmpty {
            request.setValue("Bearer \(idToken)", forHTTPHeaderField: "Authorization")
        }

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            
            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                do {
                    let decoded = try JSONDecoder().decode(AnalyticsResponse.self, from: data)
                    self.data = decoded.analytics
                    self.isLoading = false
                } catch {
                    let rawString = String(data: data, encoding: .utf8) ?? "Unreadable data"
                    self.errorMessage = "AWS sent back: \(rawString)"
                    self.isLoading = false
                }
            } else {
                let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
                self.errorMessage = "Failed to load data. Status Code: \(statusCode)"
                self.isLoading = false
            }
        } catch {
            self.errorMessage = error.localizedDescription
            self.isLoading = false
        }
    }
}

// MARK: - Main View
struct BusinessAnalyticsView: View {
    @EnvironmentObject var vm: AuthViewModel
    @StateObject private var viewModel = BusinessAnalyticsViewModel()
    
    @State private var chartMode: AnalyticsChartMode = .trend

    private let primaryColor = Color(red: 0.22, green: 0.49, blue: 1.00)

    var body: some View {
        ScrollView {
            VStack(spacing: 24) {
                if viewModel.isLoading {
                    ProgressView()
                        .scaleEffect(1.5)
                        .padding(.top, 60)
                } else if let errorMessage = viewModel.errorMessage {
                    errorStateView(message: errorMessage)
                } else if viewModel.data.isEmpty {
                    emptyStateView
                } else {
                    kpiGrid
                    
                    chartToggle
                    
                    if chartMode == .trend {
                        trendChartSection
                    } else {
                        distributionChartSection
                    }
                    
                    leaderboardSection
                }
            }
            .padding(.vertical, 20)
        }
        .background(Color(uiColor: .systemGroupedBackground).ignoresSafeArea())
        .navigationTitle("Analytics")
        .navigationBarTitleDisplayMode(.inline)
        .task {
            await viewModel.fetchAnalytics(vm: vm)
        }
        .refreshable {
            await viewModel.fetchAnalytics(vm: vm)
        }
    }
    
    // MARK: - Dashboard Components
    
    private var kpiGrid: some View {
        VStack(spacing: 12) {
            HStack {
                VStack(alignment: .leading, spacing: 4) {
                    Text("Total Views")
                        .font(.system(size: 13, weight: .bold, design: .rounded))
                        .foregroundStyle(.secondary)
                        .textCase(.uppercase)
                    
                    Text("\(viewModel.totalViews)")
                        .font(.system(size: 42, weight: .heavy, design: .rounded))
                        .foregroundStyle(.primary)
                }
                Spacer()
                Image(systemName: "eye.fill")
                    .font(.system(size: 32))
                    .foregroundStyle(primaryColor.opacity(0.2))
            }
            .padding(20)
            .background(Color(uiColor: .secondarySystemGroupedBackground))
            .clipShape(RoundedRectangle(cornerRadius: 20, style: .continuous))
            .shadow(color: .black.opacity(0.03), radius: 8, x: 0, y: 2)
            
            HStack(spacing: 12) {
                kpiCard(title: "Landmarks", value: "\(viewModel.data.count)", icon: "mappin.and.ellipse", color: .purple)
                kpiCard(title: "Top Performer", value: viewModel.topLandmark, icon: "star.fill", color: .orange)
            }
        }
        .padding(.horizontal, 20)
    }
    
    private func kpiCard(title: String, value: String, icon: String, color: Color) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Image(systemName: icon)
                    .foregroundStyle(color)
                Text(title)
                    .font(.system(size: 12, weight: .bold, design: .rounded))
                    .foregroundStyle(.secondary)
                    .textCase(.uppercase)
                Spacer()
            }
            
            Text(value)
                .font(.system(size: 18, weight: .bold, design: .rounded))
                .foregroundStyle(.primary)
                .lineLimit(1)
                .minimumScaleFactor(0.6)
        }
        .padding(16)
        .background(Color(uiColor: .secondarySystemGroupedBackground))
        .clipShape(RoundedRectangle(cornerRadius: 20, style: .continuous))
        .shadow(color: .black.opacity(0.03), radius: 8, x: 0, y: 2)
    }
    
    private var chartToggle: some View {
        Picker("Chart Mode", selection: $chartMode) {
            Text("Over Time").tag(AnalyticsChartMode.trend)
            Text("Breakdown").tag(AnalyticsChartMode.distribution)
        }
        .pickerStyle(.segmented)
        .padding(.horizontal, 20)
    }
    
    private var trendChartSection: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Engagement Over Time (30 Days)")
                .font(.system(size: 16, weight: .bold, design: .rounded))
                .padding(.horizontal, 20)
            
            VStack {
                if viewModel.totalViews == 0 {
                    noDataChartPlaceholder
                } else {
                    Chart(viewModel.timeSeriesData) { item in
                        AreaMark(
                            x: .value("Date", item.date),
                            y: .value("Views", item.clicks)
                        )
                        .foregroundStyle(
                            LinearGradient(
                                colors: [primaryColor.opacity(0.5), primaryColor.opacity(0.0)],
                                startPoint: .top,
                                endPoint: .bottom
                            )
                        )
                        
                        LineMark(
                            x: .value("Date", item.date),
                            y: .value("Views", item.clicks)
                        )
                        .foregroundStyle(primaryColor)
                        .lineStyle(StrokeStyle(lineWidth: 3, lineCap: .round, lineJoin: .round))
                    }
                    .frame(height: 220)
                    .chartXAxis {
                        AxisMarks(preset: .aligned, values: .automatic(desiredCount: 5)) { value in
                            AxisGridLine()
                            AxisValueLabel(format: .dateTime.month().day())
                        }
                    }
                }
            }
            .padding(20)
            .background(Color(uiColor: .secondarySystemGroupedBackground))
            .clipShape(RoundedRectangle(cornerRadius: 24, style: .continuous))
            .shadow(color: .black.opacity(0.04), radius: 8, x: 0, y: 4)
            .padding(.horizontal, 20)
        }
        .transition(.opacity.combined(with: .scale(scale: 0.98)))
    }
    
    private var distributionChartSection: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Views by Landmark")
                .font(.system(size: 16, weight: .bold, design: .rounded))
                .padding(.horizontal, 20)
            
            VStack {
                if viewModel.activeLandmarksWithViews.isEmpty {
                    noDataChartPlaceholder
                } else {
                    if #available(iOS 17.0, *) {
                        Chart(viewModel.activeLandmarksWithViews) { landmark in
                            SectorMark(
                                angle: .value("Views", landmark.totalClicks),
                                innerRadius: .ratio(0.618),
                                angularInset: 2.0
                            )
                            .cornerRadius(6)
                            .foregroundStyle(by: .value("Landmark", landmark.label))
                            .annotation(position: .overlay) {
                                // Only show percentage on large slices
                                let percentage = Double(landmark.totalClicks) / Double(viewModel.totalViews)
                                if percentage > 0.10 {
                                    Text("\(Int(percentage * 100))%")
                                        .font(.system(size: 12, weight: .bold, design: .rounded))
                                        .foregroundStyle(.white)
                                        .shadow(radius: 2)
                                }
                            }
                        }
                        .frame(height: 220)
                        .chartLegend(position: .bottom, alignment: .center, spacing: 16)
                    } else {
                        // Fallback Bar Chart for iOS 16
                        Chart(viewModel.activeLandmarksWithViews) { landmark in
                            BarMark(
                                x: .value("Views", landmark.totalClicks),
                                y: .value("Landmark", landmark.label)
                            )
                            .foregroundStyle(by: .value("Landmark", landmark.label))
                            .cornerRadius(4)
                        }
                        .frame(height: 220)
                        .chartLegend(.hidden)
                    }
                }
            }
            .padding(20)
            .background(Color(uiColor: .secondarySystemGroupedBackground))
            .clipShape(RoundedRectangle(cornerRadius: 24, style: .continuous))
            .shadow(color: .black.opacity(0.04), radius: 8, x: 0, y: 4)
            .padding(.horizontal, 20)
        }
        .transition(.opacity.combined(with: .scale(scale: 0.98)))
    }
    
    private var noDataChartPlaceholder: some View {
        ZStack {
            Color(uiColor: .tertiarySystemGroupedBackground)
            VStack(spacing: 8) {
                Image(systemName: "eye.slash.fill")
                    .font(.system(size: 24))
                    .foregroundStyle(.secondary)
                Text("No views recorded yet.")
                    .font(.subheadline.bold())
                    .foregroundStyle(.secondary)
            }
        }
        .frame(height: 220)
        .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
    }
    
    private var leaderboardSection: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("All Landmarks")
                .font(.system(size: 16, weight: .bold, design: .rounded))
                .padding(.horizontal, 20)
                .padding(.top, 8)
            
            VStack(spacing: 0) {
                ForEach(Array(viewModel.sortedLandmarks.enumerated()), id: \.element.id) { index, landmark in
                    HStack(spacing: 16) {
                        
                        // Rank Number
                        Text("\(index + 1)")
                            .font(.system(size: 14, weight: .bold, design: .monospaced))
                            .foregroundStyle(.tertiary)
                            .frame(width: 24, alignment: .leading)
                        
                        Text(landmark.label.isEmpty ? "Untitled Landmark" : landmark.label)
                            .font(.system(size: 16, weight: .semibold, design: .rounded))
                            .foregroundStyle(.primary)
                        
                        Spacer()
                        
                        Text("\(landmark.totalClicks)")
                            .font(.system(size: 16, weight: .bold, design: .rounded))
                            .padding(.horizontal, 12)
                            .padding(.vertical, 6)
                            .background(landmark.totalClicks > 0 ? primaryColor.opacity(0.15) : Color.gray.opacity(0.15))
                            .foregroundStyle(landmark.totalClicks > 0 ? primaryColor : .secondary)
                            .clipShape(Capsule())
                    }
                    .padding(.horizontal, 20)
                    .padding(.vertical, 14)
                    
                    if landmark.landmarkId != viewModel.sortedLandmarks.last?.landmarkId {
                        Divider().padding(.leading, 60)
                    }
                }
            }
            .background(Color(uiColor: .secondarySystemGroupedBackground))
            .clipShape(RoundedRectangle(cornerRadius: 20, style: .continuous))
            .shadow(color: .black.opacity(0.03), radius: 8, x: 0, y: 2)
            .padding(.horizontal, 20)
        }
    }
    
    // MARK: - Empty / Error States
    private func errorStateView(message: String) -> some View {
        VStack(spacing: 12) {
            Image(systemName: "exclamationmark.triangle.fill")
                .font(.system(size: 32))
                .foregroundStyle(.orange)
            Text(message)
                .font(.headline)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
        }
        .padding(.top, 40)
        .padding(.horizontal, 32)
    }
    
    private var emptyStateView: some View {
        VStack(spacing: 12) {
            Image(systemName: "chart.bar.xaxis")
                .font(.system(size: 42))
                .foregroundStyle(.secondary.opacity(0.5))
            Text("No landmarks found.")
                .font(.headline)
                .foregroundStyle(.secondary)
            Text("Add landmarks to your business account to start tracking engagement.")
                .font(.subheadline)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
        }
        .padding(.top, 60)
        .padding(.horizontal, 32)
    }
}

File: Android/LookSee/app/src/main/java/looksee/angelll/com/uifiles/BusinessAnalyticsView.kt
package looksee.angelll.com.uifiles

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.BarChart
import androidx.compose.material.icons.filled.Place
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material.icons.filled.Visibility
import androidx.compose.material.icons.filled.VisibilityOff
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.gson.Gson
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import looksee.angelll.com.models.AnalyticsResponse
import looksee.angelll.com.models.LandmarkAnalyticsData
import looksee.angelll.com.viewmodels.AuthViewModel
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*

// Internal model for Time-Series
data class TimeSeriesPoint(val date: Date, val clicks: Int)

enum class AnalyticsChartMode { Trend, Distribution }

class BusinessAnalyticsViewModel : ViewModel() {
    var data by mutableStateOf<List<LandmarkAnalyticsData>>(emptyList())
    var isLoading by mutableStateOf(true)
    var errorMessage by mutableStateOf<String?>(null)

    val totalViews: Int get() = data.sumOf { it.totalClicks }

    val topLandmark: String
        get() = if (totalViews > 0) data.maxByOrNull { it.totalClicks }?.label ?: "N/A" else "N/A"

    val activeLandmarksWithViews: List<LandmarkAnalyticsData>
        get() = data.filter { it.totalClicks > 0 }.sortedByDescending { it.totalClicks }

    val sortedLandmarks: List<LandmarkAnalyticsData>
        get() = data.sortedWith(compareByDescending<LandmarkAnalyticsData> { it.totalClicks }.thenBy { it.label })

    val timeSeriesData: List<TimeSeriesPoint>
        get() {
            val template = data.firstOrNull()?.dailyData ?: return emptyList()
            val merged = mutableMapOf<String, Int>()
            for (item in data) {
                for (point in item.dailyData) {
                    merged[point.date] = (merged[point.date] ?: 0) + point.clicks
                }
            }
            val currentYear = Calendar.getInstance().get(Calendar.YEAR)
            val formatter = SimpleDateFormat("MM/dd/yyyy", Locale.US)
            return template.mapNotNull { point ->
                val exactDate = formatter.parse("${point.date}/$currentYear") ?: return@mapNotNull null
                TimeSeriesPoint(exactDate, merged[point.date] ?: 0)
            }.sortedBy { it.date }
        }

    fun fetchAnalytics(vm: AuthViewModel) {
        if (vm.userId.isEmpty()) return
        isLoading = true
        errorMessage = null

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val token = vm.fetchIdToken()
                val url = URL("https://d11vl3v9w133rh.cloudfront.net/analytics")
                val conn = url.openConnection() as HttpURLConnection
                conn.requestMethod = "GET"
                if (token.isNotEmpty()) {
                    conn.setRequestProperty("Authorization", "Bearer $token")
                }

                val code = conn.responseCode
                val responseData = if (code in 200..299) {
                    conn.inputStream.bufferedReader().use { it.readText() }
                } else {
                    conn.errorStream?.bufferedReader()?.use { it.readText() }
                }

                withContext(Dispatchers.Main) {
                    if (code in 200..299 && responseData != null) {
                        try {
                            val decoded = Gson().fromJson(responseData, AnalyticsResponse::class.java)
                            data = decoded.analytics
                            isLoading = false
                        } catch (e: Exception) {
                            errorMessage = "AWS sent back: $responseData"
                            isLoading = false
                        }
                    } else {
                        errorMessage = "Failed to load data. Status Code: $code"
                        isLoading = false
                    }
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    errorMessage = e.localizedMessage
                    isLoading = false
                }
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun BusinessAnalyticsView(
    vm: AuthViewModel,
    onBack: () -> Unit
) {
    val viewModel = remember { BusinessAnalyticsViewModel() }
    var chartMode by remember { mutableStateOf(AnalyticsChartMode.Trend) }

    val primaryColor = Color(0xFF387DFF)
    val secondaryGrouped = Color(0xFF1C1C1E)

    LaunchedEffect(Unit) {
        viewModel.fetchAnalytics(vm)
    }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Analytics", fontWeight = FontWeight.Bold, color = Color.White) },
                colors = TopAppBarDefaults.topAppBarColors(containerColor = Color(0xFF0F0F1A)),
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back", tint = Color.White)
                    }
                }
            )
        },
        containerColor = Color(0xFF0F0F1A)
    ) { paddingValues ->
        Box(modifier = Modifier.fillMaxSize().padding(paddingValues)) {
            if (viewModel.isLoading) {
                Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                    CircularProgressIndicator(color = primaryColor)
                }
            } else if (viewModel.errorMessage != null) {
                ErrorStateView(viewModel.errorMessage!!)
            } else if (viewModel.data.isEmpty()) {
                AnalyticsEmptyStateView()
            } else {
                Column(
                    modifier = Modifier
                        .fillMaxSize()
                        .verticalScroll(rememberScrollState())
                        .padding(vertical = 20.dp),
                    verticalArrangement = Arrangement.spacedBy(24.dp)
                ) {
                    KpiGrid(viewModel, primaryColor, secondaryGrouped)

                    ChartToggle(chartMode) { chartMode = it }

                    if (chartMode == AnalyticsChartMode.Trend) {
                        TrendChartSection(viewModel, primaryColor, secondaryGrouped)
                    } else {
                        DistributionChartSection(viewModel, secondaryGrouped)
                    }

                    LeaderboardSection(viewModel, primaryColor, secondaryGrouped)
                }
            }
        }
    }
}

@Composable
private fun KpiGrid(viewModel: BusinessAnalyticsViewModel, primaryColor: Color, bg: Color) {
    Column(modifier = Modifier.padding(horizontal = 20.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
        // Total Views
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .shadow(4.dp, RoundedCornerShape(20.dp))
                .background(bg, RoundedCornerShape(20.dp))
                .padding(20.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(4.dp)) {
                Text("TOTAL VIEWS", fontSize = 13.sp, fontWeight = FontWeight.Bold, color = Color.Gray)
                Text("${viewModel.totalViews}", fontSize = 42.sp, fontWeight = FontWeight.ExtraBold, color = Color.White)
            }
            Icon(Icons.Default.Visibility, contentDescription = null, modifier = Modifier.size(32.dp), tint = primaryColor.copy(alpha = 0.2f))
        }

        Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            KpiCard(title = "LANDMARKS", value = "${viewModel.data.size}", icon = Icons.Default.Place, color = Color(0xFFAF52DE), bg = bg, modifier = Modifier.weight(1f))
            KpiCard(title = "TOP PERFORMER", value = viewModel.topLandmark, icon = Icons.Default.Star, color = Color(0xFFFF9F0A), bg = bg, modifier = Modifier.weight(1f))
        }
    }
}

@Composable
private fun KpiCard(title: String, value: String, icon: androidx.compose.ui.graphics.vector.ImageVector, color: Color, bg: Color, modifier: Modifier = Modifier) {
    Column(
        modifier = modifier
            .shadow(4.dp, RoundedCornerShape(20.dp))
            .background(bg, RoundedCornerShape(20.dp))
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(4.dp)) {
            Icon(icon, contentDescription = null, tint = color, modifier = Modifier.size(14.dp))
            Text(title, fontSize = 12.sp, fontWeight = FontWeight.Bold, color = Color.Gray)
        }
        Text(value, fontSize = 18.sp, fontWeight = FontWeight.Bold, color = Color.White, maxLines = 1)
    }
}

@Composable
private fun ChartToggle(selectedMode: AnalyticsChartMode, onModeSelect: (AnalyticsChartMode) -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 20.dp)
            .background(Color(0xFF2C2C2E), RoundedCornerShape(8.dp))
            .padding(2.dp)
    ) {
        val trendBg = if (selectedMode == AnalyticsChartMode.Trend) Color(0xFF3A3A3C) else Color.Transparent
        val distBg = if (selectedMode == AnalyticsChartMode.Distribution) Color(0xFF3A3A3C) else Color.Transparent

        Box(
            modifier = Modifier.weight(1f).clip(RoundedCornerShape(7.dp)).background(trendBg).clickable { onModeSelect(AnalyticsChartMode.Trend) }.padding(vertical = 6.dp),
            contentAlignment = Alignment.Center
        ) {
            Text("Over Time", fontSize = 13.sp, fontWeight = FontWeight.Medium, color = Color.White)
        }
        Box(
            modifier = Modifier.weight(1f).clip(RoundedCornerShape(7.dp)).background(distBg).clickable { onModeSelect(AnalyticsChartMode.Distribution) }.padding(vertical = 6.dp),
            contentAlignment = Alignment.Center
        ) {
            Text("Breakdown", fontSize = 13.sp, fontWeight = FontWeight.Medium, color = Color.White)
        }
    }
}

@Composable
private fun TrendChartSection(viewModel: BusinessAnalyticsViewModel, primaryColor: Color, bg: Color) {
    Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
        Text("Engagement Over Time (30 Days)", fontSize = 16.sp, fontWeight = FontWeight.Bold, color = Color.White, modifier = Modifier.padding(horizontal = 20.dp))
        Box(
            modifier = Modifier
                .padding(horizontal = 20.dp)
                .fillMaxWidth()
                .shadow(4.dp, RoundedCornerShape(24.dp))
                .background(bg, RoundedCornerShape(24.dp))
                .padding(20.dp)
        ) {
            if (viewModel.totalViews == 0) {
                NoDataChartPlaceholder()
            } else {
                SimpleLineChart(viewModel.timeSeriesData, primaryColor)
            }
        }
    }
}

@Composable
private fun DistributionChartSection(viewModel: BusinessAnalyticsViewModel, bg: Color) {
    Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
        Text("Views by Landmark", fontSize = 16.sp, fontWeight = FontWeight.Bold, color = Color.White, modifier = Modifier.padding(horizontal = 20.dp))
        Box(
            modifier = Modifier
                .padding(horizontal = 20.dp)
                .fillMaxWidth()
                .shadow(4.dp, RoundedCornerShape(24.dp))
                .background(bg, RoundedCornerShape(24.dp))
                .padding(20.dp)
        ) {
            if (viewModel.activeLandmarksWithViews.isEmpty()) {
                NoDataChartPlaceholder()
            } else {
                SimpleBarChart(viewModel.activeLandmarksWithViews)
            }
        }
    }
}

@Composable
private fun SimpleLineChart(data: List<TimeSeriesPoint>, color: Color) {
    if (data.isEmpty()) return
    androidx.compose.foundation.Canvas(modifier = Modifier.fillMaxWidth().height(220.dp)) {
        val maxViews = data.maxOf { it.clicks }.coerceAtLeast(1)
        val w = size.width
        val h = size.height

        val path = Path()
        val areaPath = Path()

        data.forEachIndexed { i, point ->
            val x = if (data.size > 1) (i.toFloat() / (data.size - 1)) * w else w / 2
            val y = h - (point.clicks.toFloat() / maxViews) * h
            if (i == 0) {
                path.moveTo(x, y)
                areaPath.moveTo(x, y)
            } else {
                path.lineTo(x, y)
                areaPath.lineTo(x, y)
            }
        }

        areaPath.lineTo(w, h)
        areaPath.lineTo(0f, h)
        areaPath.close()

        drawPath(areaPath, Brush.verticalGradient(listOf(color.copy(alpha = 0.5f), Color.Transparent)))
        drawPath(path, color, style = Stroke(width = 8f, cap = StrokeCap.Round, join = StrokeJoin.Round))
    }
}

@Composable
private fun SimpleBarChart(data: List<LandmarkAnalyticsData>) {
    val maxViews = data.maxOf { it.totalClicks }.coerceAtLeast(1)
    val colors = listOf(Color(0xFF387DFF), Color(0xFFAF52DE), Color(0xFFFF9F0A), Color(0xFF32D74B), Color(0xFFFF453A))

    Column(modifier = Modifier.fillMaxWidth().height(220.dp), verticalArrangement = Arrangement.SpaceEvenly) {
        data.take(5).forEachIndexed { index, landmark ->
            val fraction = landmark.totalClicks.toFloat() / maxViews.toFloat()
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                Text(landmark.label.take(15), color = Color.White, fontSize = 12.sp, modifier = Modifier.width(100.dp), maxLines = 1)
                Box(
                    modifier = Modifier.weight(1f).height(16.dp)
                ) {
                    Box(modifier = Modifier.fillMaxHeight().fillMaxWidth(fraction).background(colors[index % colors.size], RoundedCornerShape(4.dp)))
                }
                Text("${landmark.totalClicks}", color = Color.Gray, fontSize = 12.sp, modifier = Modifier.width(30.dp), textAlign = TextAlign.End)
            }
        }
    }
}

@Composable
private fun NoDataChartPlaceholder() {
    Box(
        modifier = Modifier.fillMaxWidth().height(220.dp).background(Color(0xFF2C2C2E), RoundedCornerShape(16.dp)),
        contentAlignment = Alignment.Center
    ) {
        Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Icon(Icons.Default.VisibilityOff, contentDescription = null, tint = Color.Gray, modifier = Modifier.size(24.dp))
            Text("No views recorded yet.", fontSize = 14.sp, fontWeight = FontWeight.Bold, color = Color.Gray)
        }
    }
}

@Composable
private fun LeaderboardSection(viewModel: BusinessAnalyticsViewModel, primaryColor: Color, bg: Color) {
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text("All Landmarks", fontSize = 16.sp, fontWeight = FontWeight.Bold, color = Color.White, modifier = Modifier.padding(horizontal = 20.dp, vertical = 8.dp))
        Column(
            modifier = Modifier
                .padding(horizontal = 20.dp)
                .fillMaxWidth()
                .shadow(4.dp, RoundedCornerShape(20.dp))
                .background(bg, RoundedCornerShape(20.dp))
        ) {
            viewModel.sortedLandmarks.forEachIndexed { index, landmark ->
                Row(
                    modifier = Modifier.padding(horizontal = 20.dp, vertical = 14.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    Text("${index + 1}", fontSize = 14.sp, fontWeight = FontWeight.Bold, color = Color.Gray, modifier = Modifier.width(24.dp))
                    Text(if (landmark.label.isEmpty()) "Untitled Landmark" else landmark.label, fontSize = 16.sp, fontWeight = FontWeight.SemiBold, color = Color.White, modifier = Modifier.weight(1f))
                    Box(
                        modifier = Modifier
                            .background(if (landmark.totalClicks > 0) primaryColor.copy(alpha = 0.15f) else Color.Gray.copy(alpha = 0.15f), CircleShape)
                            .padding(horizontal = 12.dp, vertical = 6.dp)
                    ) {
                        Text("${landmark.totalClicks}", fontSize = 16.sp, fontWeight = FontWeight.Bold, color = if (landmark.totalClicks > 0) primaryColor else Color.Gray)
                    }
                }
                if (index < viewModel.sortedLandmarks.size - 1) {
                    HorizontalDivider(modifier = Modifier.padding(start = 60.dp), color = Color.White.copy(alpha = 0.1f))
                }
            }
        }
    }
}

@Composable
private fun ErrorStateView(message: String) {
    Column(modifier = Modifier.fillMaxSize().padding(top = 40.dp, start = 32.dp, end = 32.dp), horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Icon(Icons.Default.Warning, contentDescription = null, tint = Color.Red, modifier = Modifier.size(32.dp))
        Text(message, fontSize = 16.sp, fontWeight = FontWeight.Bold, color = Color.Gray, textAlign = TextAlign.Center)
    }
}

@Composable
private fun AnalyticsEmptyStateView() {
    Column(modifier = Modifier.fillMaxSize().padding(top = 60.dp, start = 32.dp, end = 32.dp), horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Icon(Icons.Default.BarChart, contentDescription = null, tint = Color.Gray.copy(alpha = 0.5f), modifier = Modifier.size(42.dp))
        Text("No landmarks found.", fontSize = 16.sp, fontWeight = FontWeight.Bold, color = Color.Gray)
        Text("Add landmarks to your business account to start tracking engagement.", fontSize = 14.sp, color = Color.Gray, textAlign = TextAlign.Center)
    }
}
